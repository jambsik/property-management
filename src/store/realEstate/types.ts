export const GET_REAL_ESTATE_DATA = 'realEstate/getData';
export const GET_REAL_ESTATE_META_DATA = 'realEstate/getMetaData';
export const GET_REAL_ESTATE_ITEM = 'realEstate/getItem';
