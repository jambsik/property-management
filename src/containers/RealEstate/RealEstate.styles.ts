import { emerald } from './../../theme/colors';

export const realEstateRootStyles = {
    width: '100%',
    height: 'calc(100vh + 600px) ',
    backgroundColor: emerald,
    display: 'flex',
    flexDirection: 'column' as const,
};

export const realEstateContentStyles = {
    width: '100%',
    height: '100%',
    display: 'flex',
    flexDirection: 'column' as const,
    alignItems: 'center',
};
