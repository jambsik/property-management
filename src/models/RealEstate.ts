export interface RealEstate {
    id: number;
    mssubclass: number;
    lotfrontage: number;
    lotarea: number;
    overallqual: number;
    overallcond: number;
    yearbuilt: number;
    yearremodadd: number;
    masvnrarea: number;
    bsmtfinsf1: number;
    bsmtfinsf2: number;
    bsmtunfsf: number;
    totalbsmtsf: number;
    '1stflrsf': number;
    '2ndflrsf': number;
    lowqualfinsf: number;
    grlivarea: number;
    bsmtfullbath: number;
    bsmthalfbath: number;
    fullbath: number;
    halfbath: number;
    bedroomabvgr: number;
    kitchenabvgr: number;
    totrmsabvgrd: number;
    fireplaces: number;
    garageyrblt: number;
    garagecars: number;
    garagearea: number;
    wooddecksf: number;
    openporchsf: number;
    enclosedporch: number;
    '3ssnporch': number;
    screenporch: number;
    poolarea: number;
    miscval: number;
    mosold: number;
    yrsold: number;
    saleprice: number;
    mszoning: string;
    street: string;
    alley: string;
    lotshape: string;
    landcontour: string;
    utilities: string;
    lotconfig: string;
    landslope: string;
    neighborhood: string;
    condition1: string;
    condition2: string;
    bldgtype: string;
    housestyle: string;
    roofstyle: string;
    roofmatl: string;
    exterior1st: string;
    exterior2nd: string;
    masvnrtype: string;
    exterqual: string;
    extercond: string;
    foundation: string;
    bsmtqual: string;
    bsmtcond: string;
    bsmtexposure: string;
    bsmtfintype1: string;
    bsmtfintype2: string;
    heating: string;
    heatingqc: string;
    centralair: string;
    electrical: string;
    kitchenqual: string;
    functional: string;
    fireplacequ: string;
    garagetype: string;
    garagefinish: string;
    garagequal: string;
    garagecond: string;
    paveddrive: string;
    poolqc: string;
    fence: string;
    miscfeature: string;
    saletype: string;
    salecondition: string;
}
