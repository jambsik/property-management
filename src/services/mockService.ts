import axios from 'axios';
import { Endpoints } from './../constants/Endpoints';
import { MockModel } from './../models/MockModel';

export const API_URL = process.env.REACT_APP_MOCK_API_URL;

export const getMockData = (): Promise<MockModel | null> =>
    new Promise(async (resolve) => {
        console.log(`${API_URL}/${Endpoints.MockUrl}`);
        try {
            const { data } = await axios.get(`${API_URL}/${Endpoints.MockUrl}`, {
                headers: {},
                params: {},
            });

            resolve(data);
        } catch (error) {
            resolve(null);
        }
    });
