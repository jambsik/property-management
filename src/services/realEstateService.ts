import { Pagination } from './../constants/Pagination';
import { FilterParams, applyFilters, getNumberOfPages, findById } from './../helpers/simulateBe';
import axios from 'axios';

import { Endpoints } from './../constants/Endpoints';
import { RealEstate } from './../models/RealEstate';
import { Response } from '../models/Response';
import { applyPagination } from '../helpers/simulateBe';

export const API_URL = process.env.REACT_APP_API_URL;

export const fetchGetRealEstateData = (filters?: FilterParams, page = Pagination.DefaultPage): Promise<Response<RealEstate>> =>
    new Promise(async (resolve) => {
        try {
            const { data } = await axios.get(`${API_URL}/${Endpoints.RealEstate}`, {
                headers: {},
                params: {},
            });

            const response = data;
            const filteredData = filters ? applyFilters(filters, response.data) : response.data;
            const count = getNumberOfPages(filteredData.length);

            response.data = applyPagination<RealEstate>(filteredData, page, Pagination.DefaultLimit);

            response.metaData = {
                pagination: {
                    count,
                },
            };

            resolve(response);
        } catch (error) {
            resolve({
                data: [],
            });
        }
    });

export const fetchGetRealEstateMetaData = (): Promise<Response<void>> =>
    new Promise(async (resolve) => {
        try {
            const { data } = await axios.get(`${API_URL}/${Endpoints.RealEstateFilters}`, {
                headers: {},
                params: {},
            });
            resolve(data);
        } catch (error) {
            resolve({
                data: [],
            });
        }
    });

export const fetchGetRealEstateItemData = (id: string): Promise<RealEstate | undefined> =>
    new Promise(async (resolve) => {
        try {
            const { data } = await axios.get(`${API_URL}/${Endpoints.RealEstate}`, {
                headers: {},
                params: {},
            });

            const response = data;

            resolve(findById<RealEstate>(parseInt(id), response.data));
        } catch (error) {
            resolve(undefined);
        }
    });
