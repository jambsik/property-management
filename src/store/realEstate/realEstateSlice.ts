import { getRealStateMetaDataAction } from './actions/getRealEstateMetaData';
import { ResponseFilter } from './../../models/Response';
import { RealEstate } from './../../models/RealEstate';
import { getRealEstateDataAction } from './actions/getRealEstateData';
import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { RootState } from '../../store';
import { Response } from '../../models/Response';

export interface RealEstateSliceState {
    filters: Array<ResponseFilter>;
    items: Array<RealEstate>;
}
const initialState: RealEstateSliceState = {
    filters: [],
    items: [],
};

export const realEstateSlice = createSlice({
    name: 'realEstate',
    initialState,
    reducers: {},
    extraReducers: (builder) => {
        builder.addCase(getRealEstateDataAction.fulfilled, (state: RealEstateSliceState, action: PayloadAction<Response<RealEstate>>) => {
            state.items = action.payload.data as Array<RealEstate>;
        });

        builder.addCase(getRealStateMetaDataAction.fulfilled, (state: RealEstateSliceState, action: PayloadAction<Response<void>>) => {
            const hasFilters = action.payload.metaData?.filters;

            if (hasFilters) {
                state.filters = action.payload.metaData?.filters!;
            }
        });
    },
});

export const actions = realEstateSlice.actions;

export const selectRealEstateItems = (state: RootState) => state.realEstate.items;
export const selectRealEstateFilters = (state: RootState) => state.realEstate.filters;

export default realEstateSlice.reducer;
