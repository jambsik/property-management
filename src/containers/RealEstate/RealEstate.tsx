import { Box, styled } from '@mui/system';
import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import RealEstateFilter from '../../components/RealEstateFilter/RealEstateFilter';
import RealEstateList from '../../components/RealEstateList/RealEstateList';
import { FilterParams } from '../../helpers/simulateBe';
import { RealEstate as RealEstateModel } from '../../models/RealEstate';
import { ResponseFilter, ResponsePagination } from '../../models/Response';
import { getRealEstateDataAction } from '../../store/realEstate/actions/getRealEstateData';
import { getRealStateMetaDataAction } from '../../store/realEstate/actions/getRealEstateMetaData';
import { selectRealEstateFilters, selectRealEstateItems, selectRealEstatePagination } from '../../store/realEstate/realEstateSlice';
import { realEstateContentStyles, realEstateIconStyles, realEstateRootStyles } from './RealEstate.styles';

const RealEstate = () => {
    const Image = styled('img')``;
    const dispatch = useDispatch();
    const filters: Array<ResponseFilter> = useSelector(selectRealEstateFilters);
    const items: Array<RealEstateModel> = useSelector(selectRealEstateItems);
    const pagination: ResponsePagination | undefined = useSelector(selectRealEstatePagination);

    const getDataHandler = (filterValues?: FilterParams) => {
        dispatch(getRealEstateDataAction(filterValues));
    };

    useEffect(() => {
        dispatch(getRealStateMetaDataAction());
    }, [dispatch]);

    return (
        <Box sx={realEstateRootStyles}>
            <Image sx={realEstateIconStyles} alt="web-icon" src="images/wallet.svg" />
            <Box sx={realEstateContentStyles}>
                <RealEstateFilter filters={filters} onSearch={getDataHandler} />
                <RealEstateList pagination={pagination} items={items} onSearch={getDataHandler} />
            </Box>
        </Box>
    );
};

export default RealEstate;
