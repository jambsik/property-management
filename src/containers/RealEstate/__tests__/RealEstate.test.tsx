import React from 'react';
import * as realEstateService from '../../../services/realEstateService';
import { realEstateInitialState } from '../../../store/realEstate/realEstateSlice';
import { basicFiltersDataMock } from '../../../__tests__/__fixtures__/filtersDataMock';
import { basicRealEstateItemsDataMock } from '../../../__tests__/__fixtures__/realEstateDataMock';

import { storeRender, fireEvent } from '../../../__tests__/__utils__/reduxHelper';
import { addTestProviders } from '../../../__tests__/__utils__/renderHelper';
import RealEstate from '../RealEstate';

describe('Real Estate container test', () => {
    const getRealEstateDataActionSpy = jest.spyOn(realEstateService, 'fetchGetRealEstateData');
    const getRealStateMetaDataActionSpy = jest.spyOn(realEstateService, 'fetchGetRealEstateMetaData');

    beforeEach(() => jest.clearAllMocks());

    const pageObject = {
        streetFilter: 'input[id="street"]',
        iconButton: 'filter-box-button',
        pageTwo: 2,
    };

    beforeEach(() => jest.clearAllMocks());
    it('Should render component without errors', () => {
        const { container } = storeRender(<RealEstate />);

        expect(container).toMatchSnapshot();
    });

    it('Should get metaData filters on init', () => {
        storeRender(<RealEstate />);

        expect(getRealStateMetaDataActionSpy).toHaveBeenCalledTimes(1);
    });

    it('Should getData without filters', () => {
        const { getByTestId } = storeRender(<RealEstate />, {
            preloadedState: {
                realEstate: {
                    ...realEstateInitialState,
                    filters: basicFiltersDataMock,
                },
            },
        });

        fireEvent.click(getByTestId(pageObject.iconButton));

        expect(getRealEstateDataActionSpy).toHaveBeenCalledTimes(1);
        expect(getRealEstateDataActionSpy).toHaveBeenCalledWith({}, undefined);
    });

    it('Should getData with street filter', () => {
        const { container, getByTestId } = storeRender(<RealEstate />, {
            preloadedState: {
                realEstate: {
                    ...realEstateInitialState,
                    filters: basicFiltersDataMock,
                },
            },
        });
        const streetInput = container.querySelector(pageObject.streetFilter);

        expect(streetInput).toBeInTheDocument();

        fireEvent.change(streetInput!, {
            target: {
                value: 'Pave',
            },
        });
        fireEvent.click(getByTestId(pageObject.iconButton));

        expect(getRealEstateDataActionSpy).toHaveBeenCalledTimes(1);
        expect(getRealEstateDataActionSpy).toHaveBeenCalledWith({ street: 'Pave' }, undefined);
    });

    it('Should can change the page if click in pagination footer', () => {
        const { queryByText } = storeRender(addTestProviders(<RealEstate />), {
            preloadedState: {
                realEstate: {
                    ...realEstateInitialState,
                    items: basicRealEstateItemsDataMock,
                    pagination: {
                        count: 24,
                    },
                },
            },
        });

        const pageElement = queryByText(pageObject.pageTwo, {
            selector: 'button',
        });

        expect(pageElement).toBeInTheDocument();

        fireEvent.click(pageElement!);

        expect(getRealEstateDataActionSpy).toHaveBeenCalledTimes(1);
    });
});
