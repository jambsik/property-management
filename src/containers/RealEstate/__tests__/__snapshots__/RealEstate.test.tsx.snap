// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Real Estate container test Should render component without errors 1`] = `
<div>
  <div
    class="MuiBox-root css-1nrtrq8"
  >
    <div
      class="MuiBox-root css-ku36ua"
    >
      <div
        class="MuiBox-root css-zti9x5"
      >
        <div
          class="MuiBox-root css-1i6msmp"
        >
          <div
            class="MuiBox-root css-iryi8b"
          >
            <button
              aria-label="searcheable real estate"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorSecondary MuiIconButton-sizeMedium css-1gws2xf-MuiButtonBase-root-MuiIconButton-root"
              data-testid="filter-box-button"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge css-tzssek-MuiSvgIcon-root"
                data-testid="SearchIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation3 css-amy87t-MuiPaper-root"
      >
        <div
          class="MuiBox-root css-8dn448"
        >
          <h4
            class="MuiTypography-root MuiTypography-h4 css-jxviq7-MuiTypography-root"
          >
            realEstate.list.title
          </h4>
          <img
            alt="build"
            class="css-btig0q"
            src="/images/build.svg"
          />
          <h4
            class="MuiTypography-root MuiTypography-h4 css-jxviq7-MuiTypography-root"
          >
            realEstate.list.subTitle
          </h4>
          <button
            aria-label="searcheable real estate"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorSecondary MuiIconButton-sizeMedium css-j03kxq-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge css-tzssek-MuiSvgIcon-root"
              data-testid="SearchIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
